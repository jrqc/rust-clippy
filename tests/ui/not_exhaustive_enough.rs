#![warn(clippy::not_exhaustive_enough)]

use std::io::ErrorKind;

#[non_exhaustive]
pub enum E {
    First,
    Second,
    Third,
}

#[derive(Default)]
#[non_exhaustive]
pub struct S {
    pub a: i32,
    pub b: i32,
    pub c: i32,
}

#[derive(Default)]
#[non_exhaustive]
pub struct T(pub i32, pub i32, pub i32);

fn main() {
    //////// Enum

    let e = E::First;

    match e {
        E::First => {},
        E::Second => {},
        _ => {},
    }

    //////// Struct

    let S { a: _, b: _, .. } = S::default();

    match S::default() {
        S { a: 42, b: 21, .. } => {},
        S { a: _, b: _, .. } => {},
    }

    if let S { a: 42, b: _, .. } = S::default() {}

    let v = vec![S::default()];

    for S { a: _, b: _, .. } in v {}

    //////// Tuple Struct

    let T { 0: _, 1: _, .. } = T::default();

    match T::default() {
        T { 0: 42, 1: 21, .. } => {},
        T { 0: _, 1: _, .. } => {},
    }

    if let T { 0: 42, 1: _, .. } = T::default() {}

    let v = vec![T::default()];
    for T { 0: _, 1: _, .. } in v {}

    // Enum - Another Crate
    let error_kind = ErrorKind::ConnectionReset;

    match error_kind {
        ErrorKind::NotFound => {},
        ErrorKind::PermissionDenied => {},
        ErrorKind::ConnectionRefused => {},
        ErrorKind::ConnectionReset => {},
        ErrorKind::ConnectionAborted => {},
        ErrorKind::NotConnected => {},
        ErrorKind::AddrInUse => {},
        ErrorKind::AddrNotAvailable => {},
        ErrorKind::BrokenPipe => {},
        ErrorKind::AlreadyExists => {},
        _ => {},
    }

    // Struct - Another Crate
    // todo
}
