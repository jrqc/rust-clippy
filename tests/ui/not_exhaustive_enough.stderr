error: Enum not exhaustive enough
  --> $DIR/not_exhaustive_enough.rs:29:11
   |
LL |     match e {
   |           ^ help: try adding missing field/s: `E::Third`
   |
   = note: `-D clippy::not-exhaustive-enough` implied by `-D warnings`

error: Struct not exhaustive enough
  --> $DIR/not_exhaustive_enough.rs:37:9
   |
LL |     let S { a: _, b: _, .. } = S::default();
   |         ^^^^^^^^^^^^^^^^^^^^ help: try adding missing field/s: `c`

error: Struct not exhaustive enough
  --> $DIR/not_exhaustive_enough.rs:40:9
   |
LL |         S { a: 42, b: 21, .. } => {},
   |         ^^^^^^^^^^^^^^^^^^^^^^ help: try adding missing field/s: `c`

error: Struct not exhaustive enough
  --> $DIR/not_exhaustive_enough.rs:41:9
   |
LL |         S { a: _, b: _, .. } => {},
   |         ^^^^^^^^^^^^^^^^^^^^ help: try adding missing field/s: `c`

error: Struct not exhaustive enough
  --> $DIR/not_exhaustive_enough.rs:44:12
   |
LL |     if let S { a: 42, b: _, .. } = S::default() {}
   |            ^^^^^^^^^^^^^^^^^^^^^ help: try adding missing field/s: `c`

error: Struct not exhaustive enough
  --> $DIR/not_exhaustive_enough.rs:48:9
   |
LL |     for S { a: _, b: _, .. } in v {}
   |         ^^^^^^^^^^^^^^^^^^^^ help: try adding missing field/s: `c`

error: Struct not exhaustive enough
  --> $DIR/not_exhaustive_enough.rs:52:9
   |
LL |     let T { 0: _, 1: _, .. } = T::default();
   |         ^^^^^^^^^^^^^^^^^^^^ help: try adding missing field/s: `2`

error: Struct not exhaustive enough
  --> $DIR/not_exhaustive_enough.rs:55:9
   |
LL |         T { 0: 42, 1: 21, .. } => {},
   |         ^^^^^^^^^^^^^^^^^^^^^^ help: try adding missing field/s: `2`

error: Struct not exhaustive enough
  --> $DIR/not_exhaustive_enough.rs:56:9
   |
LL |         T { 0: _, 1: _, .. } => {},
   |         ^^^^^^^^^^^^^^^^^^^^ help: try adding missing field/s: `2`

error: Struct not exhaustive enough
  --> $DIR/not_exhaustive_enough.rs:59:12
   |
LL |     if let T { 0: 42, 1: _, .. } = T::default() {}
   |            ^^^^^^^^^^^^^^^^^^^^^ help: try adding missing field/s: `2`

error: Struct not exhaustive enough
  --> $DIR/not_exhaustive_enough.rs:62:9
   |
LL |     for T { 0: _, 1: _, .. } in v {}
   |         ^^^^^^^^^^^^^^^^^^^^ help: try adding missing field/s: `2`

error: Enum not exhaustive enough
  --> $DIR/not_exhaustive_enough.rs:67:11
   |
LL |     match error_kind {
   |           ^^^^^^^^^^ help: try adding missing field/s: `std::io::ErrorKind::WouldBlock , std::io::ErrorKind::InvalidInput , std::io::ErrorKind::InvalidData , std::io::ErrorKind::TimedOut , std::io::ErrorKind::WriteZero , std::io::ErrorKind::Interrupted , std::io::ErrorKind::Other , std::io::ErrorKind::UnexpectedEof`

error: aborting due to 12 previous errors

